aten/src/ATen/core/ivalue_inl.h

```diff
@@ -623,31 +660,54 @@ inline IValue::IValue(c10::intrusive_ptr<ivalue::Future> v)
   payload.as_intrusive_ptr = v.release();
 }
 
-inline const c10::ListPtr<int64_t>& IValue::toIntListRef() const {
+inline const c10::ListPtr<int64_t>& IValue::toIntListRef() const & {
   return toIntList()->elements();
 }
 
-inline const c10::ListPtr<double>& IValue::toDoubleListRef() const {
+inline c10::ListPtr<int64_t> IValue::toIntListRef() && {
+  return std::move(*this).toIntList()->elements();
```

Comment:
I am really confused by all the different list type here.

There is at least:
1. std::vector<T>, the standard C++ list.
2. std::intrusive_ptr<List<int|tensor|double|bool|IValue>, what is held in an ivalue
3. c10::ListPtr<int64_t>. Is this a wrapper around 2? Or does 2 hold 3?, Does that add an unnecessary indirection?

Reply:
std::vector and c10::ListPtr will stay. I'm currently working on another diff that would streamline this, remove the double indirection IValue->ivalue::List->c10::ListPtr in favor of directly storing c10::ListPtr in IValue, and if possible, remove the intrusive_ptr versions from the interface. That is unrelated to this PR though.